pkg/proxy/proxy.go

```diff
@@ -0,0 +1,218 @@
+package proxy
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/registration-service/pkg/application"
+	"github.com/codeready-toolchain/registration-service/pkg/auth"
+	"github.com/codeready-toolchain/registration-service/pkg/configuration"
+	"github.com/codeready-toolchain/registration-service/pkg/context"
+	crterrors "github.com/codeready-toolchain/registration-service/pkg/errors"
+	"github.com/codeready-toolchain/registration-service/pkg/log"
+	"github.com/codeready-toolchain/registration-service/pkg/proxy/namespace"
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	controllerlog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	ProxyPort = "8081"
+)
+
+type Proxy struct {
+	namespaces  *UserNamespaces
+	tokenParser *auth.TokenParser
+}
+
+func NewProxy(app application.Application, config configuration.RegistrationServiceConfig) (*Proxy, error) {
+	cl, err := newClusterClient()
+	if err != nil {
+		return nil, err
+	}
+	return newProxyWithClusterClient(app, config, cl)
+}
+
+func newProxyWithClusterClient(app application.Application, config configuration.RegistrationServiceConfig, cln client.Client) (*Proxy, error) {
+	// Initiate toolchain cluster cache service
+	cacheLog := controllerlog.Log.WithName("registration-service")
+	cluster.NewToolchainClusterService(cln, cacheLog, config.Namespace(), 5*time.Second)
+
+	tokenParserInstance, err := auth.DefaultTokenParser()
+	if err != nil {
+		return nil, err
+	}
+	return &Proxy{
+		namespaces:  NewUserNamespaces(app),
```

Comment:
instead of using a struct with a single method, could we refactor this `namespaces` field to a func such as:
```
type Proxy struct {
	getNamespaces  func(ctx *gin.Context, userID string) (*namespace.Namespace, error)
	tokenParser *auth.TokenParser
}
```

so we could do something like
```
p.getNamespace(ctx, userID)
```
instead of 
```
p.namespaces.GetNamespace(ctx, userID)
```
in line 116

Reply:
We could but I think it's just matter of personal preferences and not that critical in this case so I would keep it as is for now if you don't mind :)

Reply:
> @xcoulon @rajivnathan thanks for the review! @xcoulon I've addressed or commented on all your comments so please review.

thanks for taking my comments into account @alexeykazakov